kapitan/remoteinventory/fetch.py

```diff
@@ -0,0 +1,101 @@
+import os
+import logging
+from collections import defaultdict
+from functools import partial
+import hashlib
+
+from kapitan import cached
+from kapitan.dependency_manager.base import fetch_git_dependency, fetch_http_dependency
+from kapitan.utils import normalise_join_path
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_inventories(inventory_path, target_objs, save_dir, force, pool):
+    """Parses through the 'inventory' parameter in target_objs to fetch the remote
+       inventories and stores it in save_dir before recursively
+       copying it to the output_path (relative to the inventory path)
+       Overwrites existing inventory items if force fetched
+
+       :param inventory_path: default or user specified inventory path
+       :param target_objs: target objects
+       :param save_dir: directory to save the fetched items
+       :param force: bool value
+       :param pool: pool object for multiprocessing
+       :return: None
+    """
+    git_inventories = defaultdict(list)
+    http_inventories = defaultdict(list)
+    # To ensure no duplicate output path
+    inv_output_path = defaultdict(set)
+
+    for target_obj in target_objs:
+        try:
+            inventories = target_obj["inventories"]
+            for inv in inventories:
+                inv_type = inv["type"]
+                source_uri = inv["source"]
+                source_hash = hashlib.sha256(source_uri.encode())
+                # hashing the source, subdir and ref together for git sources
+                # as different inventory items can have the same git uri
+                if "subdir" in inv:
+                    subdir = inv["subdir"]
+                    source_hash.update(subdir.encode())
+                if "ref" in inv:
+                    ref = inv["ref"]
+                    source_hash = hashlib.sha256(ref.encode())
+                if source_hash in cached.inv_sources:
+                    continue
+                # output_path is relative to inventory_path
+                # ./inventory by default
+                output_path = normalise_join_path(inventory_path, inv["output_path"])
+                logger.debug("Updated output_path from {} to {}".format(inv["output_path"], output_path))
+                inv["output_path"] = output_path
+
+                if output_path in inv_output_path[source_uri]:
+                    # if the output_path is duplicated for the same source_uri
+                    logger.warning("Skipping duplicate output path for uri {}".format(source_uri))
+                    continue
+                else:
+                    inv_output_path[source_uri].add(output_path)
+
+                if inv_type == "git":
+                    git_inventories[source_uri].append(inv)
+                elif inv_type in ("http", "https"):
+                    http_inventories[source_uri].append(inv)
```

Comment:
It might be good to have an "else:" condition here in case the user specifies any inv_type different than git/http/https just print out a logger.warning message. To catch people making typos. :)

Reply:
I think a typo will be caught during schema verification while loading target inventory. Regardless, I'll add the else condition just to be safe :) https://github.com/deepmind/kapitan/blob/2fce9462e83a2a6e01ee674c1a6b7766bb95bf5a/kapitan/targets.py#L72